src/formats/pnm.c

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: MIT
+// PNM formats decoder
+// Copyright (C) 2023 Abe Wieland <abe.wieland@gmail.com>
+
+#include "loader.h"
+
+#include <limits.h>
+
+// Divide a by b, rounding to the nearest integer; evaluates b twice
+#define div_near(a, b) (((a) + (b) / 2) / (b))
+// Divide a by b, rounding up; evalutes b twice
+#define div_ceil(a, b) (((a) + (b)-1) / (b))
+
+// Although the specification states comments may also appear in integers, this
+// is not supported by any known parsers at the time of writing; comments are
+// also supposedly allowed after the final integer in the header for raw PNM
+// files, but this is again not really supported. Thus, we don't support either
+// feature for consistency with other parsers
+
+// PNM file types
+#define PNM_PBM 0
+#define PNM_PGM 1
+#define PNM_PPM 2
+
+// A file-like abstraction for cleaner number parsing
+struct pnm_iter {
+    const uint8_t* pos;
+    const uint8_t* end;
+};
+
+// Error conditions
+#define PNM_EEOF -1
+#define PNM_ERNG -2
+#define PNM_EFMT -3
+#define PNM_EOVF -4
+
+static const char* pnm_strerror(int err)
+{
+    switch (err) {
+        case PNM_EEOF:
+            return "unexpected end of image";
+        case PNM_ERNG:
+            return "integer too large";
+        case PNM_EFMT:
+            return "digit expected";
+        case PNM_EOVF:
+            return "pixel value above maxval";
+        default:
+            return "unknown error";
+    }
+}
+
+// Read an integer, ignoring leading whitespace and comments
+static int pnm_readint(struct pnm_iter* it)
+{
+    for (; it->pos != it->end; ++it->pos) {
+        if (*it->pos == '#') {
+            while (it->pos != it->end && *it->pos != '\n' && *it->pos != '\r')
+                ++it->pos;
+        } else if (*it->pos != ' ' && *it->pos != '\t' && *it->pos != '\n' &&
+                   *it->pos != '\r') {
+            // XXX isspace is still on the table, but every parser I read
+            // handled this themselves
+            break;
+        }
+    }
+    if (it->pos == it->end)
+        return PNM_EEOF;
+    if (*it->pos < '0' || *it->pos > '9')
+        return PNM_EFMT;
+    int val = 0;
+    // XXX strtol is still on the table, but to do so we'd need to make a first
+    // pass to copy to a buffer, so it seems easier to stick with the majority
+    // and do this ourselves as well
+    do {
+        uint8_t d = *it->pos - '0';
+        // XXX the standard doesn't mention how large integers may be; most of
+        // the parsers work with signed integers only
+        if (val > INT_MAX / 10)
+            return PNM_ERNG;
+        val *= 10;
+        if (val > INT_MAX - d)
+            return PNM_ERNG;
+        val += d;
+        ++it->pos;
+    } while (it->pos != it->end && *it->pos >= '0' && *it->pos <= '9');
+    return val;
+}
+
+// Decode a plain/ASCII PNM file; returns 0 on success, an error code otherwise
+static int decode_plain(struct image_frame* f, struct pnm_iter* it, int type,
```

Comment:
It would be better to use `enum` instead of `int` for the type. It will give us a type check in compile time.